src/promise.h

```diff
@@ -1,28 +1,96 @@
 #pragma once
 
 #include <kj/async.h>
+#include <kj/debug.h>
 #include <rust/cxx.h>
 
+// These types are shared with rust
 namespace kj_rs {
 
 using OwnPromiseNode = kj::_::OwnPromiseNode;
 
-void own_promise_node_drop_in_place(OwnPromiseNode*);
-
-// https://github.com/dtolnay/cxx/blob/86cd652c06c5cb4c2e24d3ab555cf707b4ae0883/src/cxx.cc#L518
 namespace repr {
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wreturn-type-c-linkage"
+
+// ::cxx::private::PtrLen
 struct PtrLen final {
-  void *ptr;
-  std::size_t len;
+  void *ptr = nullptr;
+  std::size_t len = 0;
+};
+
+extern "C" {
+repr::PtrLen cxxbridge1$exception(const char *, std::size_t len) noexcept;
+}
+
+// ::cxx::private::Result
+struct Result final {
+  PtrLen err = {};
+  inline static Result ok() { return {}; }
+  inline static Result error(kj::Exception &e);
 };
+
+// ::kj::promise::UnwrapCallback
+using UnwrapCallback = Result (*)(const void /* kj::_::PromiseNode */ *node,
+                                  void /* T */ *ret);
+// ::kj::promise::KjPromiseNodeImpl
+struct KjPromiseNodeImpl {
+  template <typename T> inline KjPromiseNodeImpl(kj::Promise<T> &&p);
+
+  kj::_::PromiseNode *node;
+  repr::UnwrapCallback unwrap = nullptr;
+};
+
+#pragma GCC diagnostic pop
+} // namespace repr
+
+namespace _ {
+template <typename T>
+static repr::Result unwrapCallback(const void *nodePtr, void *ret) noexcept {
+  auto node = OwnPromiseNode(
+      reinterpret_cast<kj::_::PromiseNode *>(const_cast<void *>(nodePtr)));
+
+  kj::_::ExceptionOr<kj::_::FixVoid<T>> result;
+  node->get(result);
+
+  // destroy the promise and return error if it fails
+  KJ_IF_SOME(e, kj::runCatchingExceptions([&node]() { node = nullptr; })) {
+    return repr::Result::error(e);
+  }
+
+  KJ_IF_SOME(e, result.exception) { return repr::Result::error(e); }
+
+  if constexpr (!kj::isSameType<T, void>()) {
+    *reinterpret_cast<T *>(ret) = KJ_ASSERT_NOT_NULL(kj::mv(result.value));
```

Comment:
I'd recommend using placement-new instead. That's how my original version works: https://github.com/capnproto/kj-rs/blob/9fc9de649bbf9832d343cebc1c6cb84f47191f77/examples/kj-rs-demo/src/promise-boilerplate.c%2B%2B#L34-L36

I think it may actually be important, because we don't expect `ret` to be a fully-initialized at this point, I think, right? So, we shouldn't invoke `operator=()` on it, which might bug out if it sees an unitialized `T` whose memory is not already a valid moved-from state.

Reply:
how's the structure now? I also tried to protect against exception during move

Reply:
LGTM!

Good thought about exceptions during move-construction... though we probably don't need to worry about them for the simple reason that if a type's move constructor throws, it is probably not safe to `memcpy()`, and thus not safe for Rust to own. (No need to change anything, though.)